File: 45-jump-game-ii/45-jump-game-ii.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int mn = 0,mx = 0;
        int n = nums.size();
        
        int step = 0,index = 0;
        while(mx < (n-1)) {
            
            for(int i=index;i<n;i++) {
                if((i + nums[i] > mx) && i <= mn) {
                    mx = i + nums[i];
                    index = i;
                }
                if(i > mn) break;
            }
            
            step++;
            mn = mx;
        }
        return step;
    }
};

File: 45-jump-game-ii/45-jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
        
        int n = nums.length;
        int max_curr = 0,max_so_far = 0;
        int jump = 0;
        for(int i=0;i<n-1;i++){
            max_so_far = Math.max(max_so_far,i + nums[i]);
            if(i == max_curr){
                jump += 1;
                max_curr = max_so_far;
            }
        }
        return jump;
    }
}